common/crypto/openssl/init.c

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include "init.h"
+#include <openssl/err.h>
+#include <openssl/pem.h>
+
+#ifdef OE_BUILD_ENCLAVE
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+#include <openssl/engine.h>
+static oe_once_t _openssl_initialize_once;
+static ENGINE* eng;
+#else
+#include <pthread.h>
+static pthread_once_t _once = PTHREAD_ONCE_INIT;
+#endif
+
+#ifdef OE_BUILD_ENCLAVE
+static void _finialize(void)
+{
+    if (eng)
+    {
+        ENGINE_finish(eng);
+        ENGINE_free(eng);
+        ENGINE_cleanup();
+    }
+}
+
+static void _initialize_rdrand_engine()
+{
+    int rc = 0;
+
+    /* Initialize rdrand engine. */
+    ENGINE_load_rdrand();
+    eng = ENGINE_by_id("rdrand");
+    if (eng == NULL)
+        goto done;
+
+    rc = ENGINE_init(eng);
+    if (rc == 0)
+        goto done;
+
+    rc = ENGINE_set_default(eng, ENGINE_METHOD_RAND);
+    if (rc == 0)
+        goto done;
+
+    if (!atexit(_finialize))
+        goto done;
+
+    rc = 1;
+
+done:
+    if (rc == 0)
+    {
+        OE_TRACE_ERROR("OpenSSL initialization failed");
+        _finialize();
+    }
+    return;
+}
+#endif
+
+static void _initialize(void)
+{
+    OpenSSL_add_all_algorithms();
```

Comment:
Can the user initialize OpenSSL differently that what we are doing?
- If yes, can OpenSSL handle double initialization?
- On the host side, presumably, the user may want to initialize OpenSSL differently. In that case, should we enforce a specific kind of initialization on the user?